### PR TITLE
[CVP] Canonicalize signed minmax into unsigned

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch turns signed minmax to unsigned to match the behavior for signed icmps.
Alive2: https://alive2.llvm.org/ce/z/UAAM42

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 1]
```diff
@@ -528,17 +530,35 @@ static bool processAbsIntrinsic(IntrinsicInst *II, LazyValueInfo *LVI) {
 }
 
 // See if this min/max intrinsic always picks it's one specific operand.
+// If not, check whether we can canonicalize signed minmax into unsigned version
 static bool processMinMaxIntrinsic(MinMaxIntrinsic *MM, LazyValueInfo *LVI) {
   CmpInst::Predicate Pred = CmpInst::getNonStrictPredicate(MM->getPredicate());
   LazyValueInfo::Tristate Result = LVI->getPredicateAt(
       Pred, MM->getLHS(), MM->getRHS(), MM, /*UseBlockValue=*/true);
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:537]: Can you please also replace this getPredicateAt() use with comparison on getConstantRangeAtUse()? It makes very little sense to do both. (getPredicateAt can be better in rare cases involving equality predicates, but it shouldn't be relevant here.)

File: llvm/test/Transforms/CorrelatedValuePropagation/min-max.ll [hunk 2]
```diff
@@ -1,5 +1,6 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
-; RUN: opt < %s -passes=correlated-propagation -S | FileCheck %s
+; RUN: opt < %s -passes=correlated-propagation -S | FileCheck %s --check-prefixes=CHECK,CHECK-CANONICALIZE-ON
+; RUN: opt < %s -passes=correlated-propagation -canonicalize-icmp-predicates-to-unsigned=false -S | FileCheck %s --check-prefixes=CHECK,CHECK-CANONICALIZE-OFF
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/min-max.ll:3]: Please remove the `-canonicalize-icmp-predicates-to-unsigned=false` check lines, we don't care about this. Feel free to remove the option in a followup.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 3]
```diff
@@ -47,11 +47,6 @@ using namespace llvm;
 
 #define DEBUG_TYPE "correlated-value-propagation"
 
-static cl::opt<bool> CanonicalizeICmpPredicatesToUnsigned(
```
Comment (danilaml) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:50]: Did something change for the example mentioned in https://reviews.llvm.org/D112895#3149487 to warrant removal of this option?

Comment (nikic) [reply llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:50]: The example is optimized by ConstraintElimination now.

**Pull request merged**